File: algorithmic-thinking/backtracking/permutation/Permutation.java
import java.util.*;

/*
使用回溯生成排列 (Java Implementation)

算法说明:
- 使用回溯生成列表的所有排列
- 使用 ArrayList 实现动态大小调整和灵活性
- 在每个递归级别，尝试每个未使用的元素
- 通过从当前排列中移除元素来回溯

时间复杂度：O(n! * n) - n! 个排列，每个需要 O(n) 时间复制
空间复杂度：O(n) - 递归深度（不计算输出）

示例:
    permute([1, 2, 3]) 返回 [[1,2,3], [1,3,2], [2,1,3], ...]
*/

public class Permutation {

    /*
     * 使用回溯生成列表的所有排列。
     *
     * @param nums 要排列的整数列表
     * @return 所有排列的列表
     */
    public static List<List<Integer>> permute(List<Integer> nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(result, new ArrayList<>(), nums);
        return result;
    }

    /*
     * 回溯的辅助方法。
     *
     * @param result 存储所有排列的列表
     * @param current 正在构建的当前排列
     * @param remaining 未使用的元素
     */
    private static void backtrack(List<List<Integer>> result,
                                   List<Integer> current,
                                   List<Integer> remaining) {
        // 基础情况：所有元素都已使用，添加到结果中
        if (remaining.isEmpty()) {
            result.add(new ArrayList<>(current));
            return;
        }

        // 尝试每个剩余元素作为下一个元素
        for (int i = 0; i < remaining.size(); i++) {
            // 选择：将 remaining.get(i) 添加到当前排列
            int num = remaining.get(i);
            current.add(num);

            // 探索：递归构建排列的其余部分
            List<Integer> newRemaining = new ArrayList<>(remaining);
            newRemaining.remove(i);
            backtrack(result, current, newRemaining);

            // 撤销：移除元素以进行回溯
            current.remove(current.size() - 1);
        }
    }

    /*
     * 使用数组原位置交换的优化排列。
     *
     * 时间复杂度：O(n! * n)
     * 空间复杂度：O(n)
     */
    public static List<List<Integer>> permuteOptimized(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrackSwap(result, nums, 0);
        return result;
    }

    /*
     * 原位置交换回溯的辅助方法。
     *
     * @param result 存储所有排列的列表
     * @param nums 要排列的数组
     * @param index 当前排列的位置
     */
    private static void backtrackSwap(List<List<Integer>> result, int[] nums, int index) {
        // 基础情况：到了数组的末举
        if (index == nums.length) {
            List<Integer> perm = new ArrayList<>();
            for (int num : nums) {
                perm.add(num);
            }
            result.add(perm);
            return;
        }

        // 尝试从 index 开始的每个元素作为下一个元素
        for (int i = index; i < nums.length; i++) {
            // 选择：交换元素
            int temp = nums[index];
            nums[index] = nums[i];
            nums[i] = temp;

            // 探索：排列余下部分
            backtrackSwap(result, nums, index + 1);

            // 撤销：交换回师
            temp = nums[index];
            nums[index] = nums[i];
            nums[i] = temp;
        }
    }

    /*
     * 排列生成的测试用例
     */
    public static void main(String[] args) {
        System.out.println("=== 排列回溯测试用例 ===");

        // 测试用例 1: [1, 2, 3]
        System.out.println("测试 1: permute([1, 2, 3])");
        List<Integer> test1 = Arrays.asList(1, 2, 3);
        List<List<Integer>> result1 = permute(test1);
        System.out.println("结果（共 " + result1.size() + " 个）:");
        for (List<Integer> perm : result1) {
            System.out.println("  " + perm);
        }
        System.out.println();

        // 测试用例 2: [1, 2]
        System.out.println("测试 2: permute([1, 2])");
        List<Integer> test2 = Arrays.asList(1, 2);
        List<List<Integer>> result2 = permute(test2);
        System.out.println("结果（共 " + result2.size() + " 个）:");
        for (List<Integer> perm : result2) {
            System.out.println("  " + perm);
        }
        System.out.println();

        // 测试用例 3: [1]
        System.out.println("测试 3: permute([1])");
        List<Integer> test3 = Arrays.asList(1);
        List<List<Integer>> result3 = permute(test3);
        System.out.println("结果（共 " + result3.size() + " 个）:");
        for (List<Integer> perm : result3) {
            System.out.println("  " + perm);
        }
        System.out.println();

        // 测试用例 4: 新优化版本 [1, 2, 3, 4]
        System.out.println("测试 4: permuteOptimized([1, 2, 3, 4])");
        int[] test4 = {1, 2, 3, 4};
        List<List<Integer>> result4 = permuteOptimized(test4);
        System.out.println("结果（共 " + result4.size() + " 个）:");
        System.out.println("前 3 个排列:");
        for (int i = 0; i < 3; i++) {
            System.out.println("  " + result4.get(i));
        }
        System.out.println("后 3 个排列:");
        for (int i = result4.size() - 3; i < result4.size(); i++) {
            System.out.println("  " + result4.get(i));
        }
        System.out.println();

        // 测试用例 5: 优化版本
        System.out.println("测试 5: permuteOptimized([10, 20, 30])");
        int[] test5 = {10, 20, 30};
        List<List<Integer>> result5 = permuteOptimized(test5);
        System.out.println("结果（共 " + result5.size() + " 个）:");
        for (List<Integer> perm : result5) {
            System.out.println("  " + perm);
        }
    }
}

File: algorithmic-thinking/backtracking/permutation/permutation.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
使用回溯生成排列 (C Implementation)

算法说明:
- 使用回溯生成数组的所有排列
- 使用原地交换以避免过度的内存分配
- 在每个递归级别，尝试交换每个剩余元素
- 通过交换回原位置来回溯

时间复杂度：O(n! * n) - n! 个排列，每个需要 O(n) 时间复制
空间复杂度：O(n) - 递归深度（不计算输出）

示例：
    permute([1, 2, 3]) 返回所有 6 个排列：
    [1,2,3], [1,3,2], [2,1,3], [2,3,1], [3,1,2], [3,2,1]
*/

typedef struct {
    int* data;
    int size;
} Permutation;

typedef struct {
    Permutation* perms;
    int count;
    int capacity;
} PermutationResult;

void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void add_permutation(PermutationResult* result, int* arr, int n) {
    if (result->count >= result->capacity) {
        result->capacity *= 2;
        result->perms = (Permutation*)realloc(result->perms,
                                               result->capacity * sizeof(Permutation));
    }

    result->perms[result->count].data = (int*)malloc(n * sizeof(int));
    memcpy(result->perms[result->count].data, arr, n * sizeof(int));
    result->perms[result->count].size = n;
    result->count++;
}

void backtrack(int* arr, int n, int index, PermutationResult* result) {
    // 基础情况：到达数组末尾
    if (index == n) {
        add_permutation(result, arr, n);
        return;
    }

    // 尝试从 index 开始的每个元素作为下一个元素
    for (int i = index; i < n; i++) {
        // 选择：交换元素
        swap(&arr[index], &arr[i]);

        // 探索：排列其余部分
        backtrack(arr, n, index + 1, result);

        // 撤销：交换回去
        swap(&arr[index], &arr[i]);
    }
}

PermutationResult* permute(int* nums, int n) {
    /*
生成 nums 数组的所有排列。

    参数:
        nums: 整数数组
        n: 数组长度

    返回:
        包含所有排列的 PermutationResult 结构体
*/
    PermutationResult* result = (PermutationResult*)malloc(sizeof(PermutationResult));
    result->capacity = 1;
    for (int i = 1; i <= n; i++) {
        result->capacity *= i;  // 容量 = n!
    }
    result->perms = (Permutation*)malloc(result->capacity * sizeof(Permutation));
    result->count = 0;

    int* arr = (int*)malloc(n * sizeof(int));
    memcpy(arr, nums, n * sizeof(int));

    backtrack(arr, n, 0, result);

    free(arr);
    return result;
}

void free_result(PermutationResult* result) {
    for (int i = 0; i < result->count; i++) {
        free(result->perms[i].data);
    }
    free(result->perms);
    free(result);
}

void print_permutation(Permutation* perm) {
    printf("[");
    for (int i = 0; i < perm->size; i++) {
        printf("%d", perm->data[i]);
        if (i < perm->size - 1) printf(", ");
    }
    printf("]");
}

int main() {
    printf("=== 排列回溯测试用例 ===\n\n");

    // 测试用例 1: [1, 2, 3]
    printf("测试 1: permute([1, 2, 3])\n");
    int test1[] = {1, 2, 3};
    PermutationResult* result1 = permute(test1, 3);
    printf("结果（共 %d 个）：\n", result1->count);
    for (int i = 0; i < result1->count; i++) {
        printf("  ");
        print_permutation(&result1->perms[i]);
        printf("\n");
    }
    free_result(result1);
    printf("\n");

    // 测试用例 2: [1, 2]
    printf("测试 2: permute([1, 2])\n");
    int test2[] = {1, 2};
    PermutationResult* result2 = permute(test2, 2);
    printf("结果（共 %d 个）：\n", result2->count);
    for (int i = 0; i < result2->count; i++) {
        printf("  ");
        print_permutation(&result2->perms[i]);
        printf("\n");
    }
    free_result(result2);
    printf("\n");

    // 测试用例 3: [1]
    printf("测试 3: permute([1])\n");
    int test3[] = {1};
    PermutationResult* result3 = permute(test3, 1);
    printf("结果（共 %d 个）：\n", result3->count);
    for (int i = 0; i < result3->count; i++) {
        printf("  ");
        print_permutation(&result3->perms[i]);
        printf("\n");
    }
    free_result(result3);
    printf("\n");

    // 测试用例 4: [10, 20, 30]
    printf("测试 4: permute([10, 20, 30])\n");
    int test4[] = {10, 20, 30};
    PermutationResult* result4 = permute(test4, 3);
    printf("结果（共 %d 个）：\n", result4->count);
    for (int i = 0; i < result4->count; i++) {
        printf("  ");
        print_permutation(&result4->perms[i]);
        printf("\n");
    }
    free_result(result4);
    printf("\n");

    // 测试用例 5: [1, 2, 3, 4]（显示前 6 个）
    printf("测试 5: permute([1, 2, 3, 4]) - 显示前 6 个\n");
    int test5[] = {1, 2, 3, 4};
    PermutationResult* result5 = permute(test5, 4);
    printf("结果（共 %d 个）：\n", result5->count);
    for (int i = 0; i < 6 && i < result5->count; i++) {
        printf("  ");
        print_permutation(&result5->perms[i]);
        printf("\n");
    }
    printf("  ... （显示 6 个，共 %d 个）\n", result5->count);
    free_result(result5);

    return 0;
}
